tensorrt_llm/_torch/modules/attention.py

```diff
@@ -126,24 +155,41 @@ def forward(
         attention_mask: PredefinedAttentionMask = PredefinedAttentionMask.
         CAUSAL,
         mrope_config: Optional[dict] = None,
+        all_reduce_params: Optional[AllReduceParams] = None,
         **kwargs,
     ) -> torch.Tensor:
         qkv = self.qkv_proj(hidden_states)
         is_fused_qkv = False
         if isinstance(self.attn, TrtllmAttention):
             is_fused_qkv = True
 
+        if self.qk_norm is not None:
+            # TODO: make this more efficient.
+            q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size],
+                                dim=-1)
+            do_multi_stream = torch.cuda.is_current_stream_capturing(
+            ) and self.aux_stream is not None
+            if do_multi_stream:
+                self.ln_events[0].record()
+                k = self.qk_norm(k)
+                with torch.cuda.stream(self.aux_stream):
+                    self.ln_events[0].wait()
+                    q = self.qk_norm(q)
+                    self.ln_events[1].record()
+                self.ln_events[1].wait()
+            else:
+                q = self.qk_norm(q)
+                k = self.qk_norm(k)
+            qkv = torch.concat([q, k, v], dim=-1)
+
         if is_fused_qkv:
             if self.pos_embd_params is None and position_ids is not None:
                 q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size],
                                     dim=-1)
                 q, k = self.rotary_emb(
                     position_ids,
                     [q.contiguous(), k.contiguous()], attn_metadata)
```

Comment:
Is this part no-op then if we're invoking the trtllm attention? If so, remove the `.contiguous()` calls.

Reply:
We should add a comment on which attention backend corresponds to which code path to make the code easier to follow.

Reply:
Leaving commenting to @yuxianq when he does his refactor